### PR TITLE
change dtype promotion behavior for jit-invariance

### LINKED ISSUE #732: Inconsistent casting of floats with jit compilation.
```
def mul(R, c):
  return R * c

R = np.array([1, 2, 3], dtype=np.float32)
c = 0.1

print(mul(R, c).dtype) # Evaluates to np.float32 as expected.
print(jit(mul)(R, c).dtype) # Evaluates to np.float64.
```

### PR DESCRIPTION
**Pull request by mattjj**
Here are two desiderata for jax.numpy dtype promotion behavior:
1. follow what NumPy does (or a cleaned-up version of it)
2. be invariant to `@jit`

The latter is much more important, so whenever the two are in tension we prefer the latter.

Issue #732 showed our code had a special behavior that essentially handled a case of the former desideratum but also broke the latter. #732 also showed us (again) that our tests really should cover Python scalars.

In summary, in this commit:
* revise jax.numpy dtype promotion behavior to be invariant to `@jit`, but not to follow NumPy behavior with Python scalars involved
* add Python scalar types to lax_numpy tests
* simplify and update kron implementation to fix dtype issues
* fix #732

### CODE REVIEW

File: jax/lax/lax.py [hunk 1]
```diff
@@ -735,7 +735,8 @@ def _get_min_identity(dtype):
     return onp.array(True, onp.bool_)
 
 def _reduce_sum(operand, axes):
-  return reduce_sum_p.bind(operand, axes=tuple(axes), input_shape=operand.shape)
+  return reduce_sum_p.bind(operand, axes=tuple(axes),
+                           input_shape=onp.shape(operand))
```
Comment (hawkinsp) [inline jax/lax/lax.py]: It's surprising to me to call onp.shape() here. (I'm unclear what onp.shape will do when passed one of our arrays.)

Perhaps define a _shape() function that delegates to onp.shape together with a comment explaining why it works and is needed?

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -1338,9 +1333,7 @@ def array_equal(a1, a2):
     a1, a2 = asarray(a1), asarray(a2)
   except Exception:
     return False
-  if a1.shape != a2.shape:
-    return False
-  return asarray(a1==a2).all()
+  return shape(a1) == shape(a2) and all(asarray(a1 == a2))
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: Is calling all(...) correct for traced values?

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -231,10 +225,20 @@ def op_record(name, nargs, dtypes, shapes, rng, diff_modes, test_name=None,
     # op_record("__xor__", 2, number_dtypes, all_shapes, jtu.rand_bool(), []),
     # op_record("__rxor__", 2, number_dtypes, all_shapes, jtu.rand_bool(), []),
     # op_record("__divmod__", 2, number_dtypes, all_shapes, jtu.rand_nonzero(), []),
-    # op_record("__rdivmod__", 2, number_dtypes, all_shapes, jtu.rand_nonzero(), []),
+    # # op_record("__rdivmod__", 2, number_dtypes, all_shapes, jtu.rand_nonzero(), []),
```
Comment (hawkinsp) [inline tests/lax_numpy_test.py]: Extra comment character.

**Review by hawkinsp — APPROVED** [review]
The only other thing I would suggest is adding a section to the documentation explaining what the new dtype promotion semantics are.

File: jax/lax/lax.py (hunk 1 above)
Comment (mattjj) [reply jax/lax/lax.py]: Actually we call onp.shape on DeviceArrays and Tracers all the time; if that didn’t do the right thing, we wouldn’t be doing a good job of emulating numpy! (All our _shape functions are just rebindings of onp.shape anyway, to make the name shorter.) All onp.shape does to arraylikes is read their “shape” attr, but it also handles scalars. 

If this were the first appearance of onp.shape used this way it might warrant a comment.

It’s needed because scalars can be passed in here, and scalars can be passed in here because per #121 jax.numpy.array(scalar) returns a scalar (rather than an arraylike with a shape attribute). That is, we don’t have a convenient way to promote scalars to 0D arrays until we fix #121.

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (mattjj) [reply jax/numpy/lax_numpy.py]: Yes it should be, because that’s jax.numpy.all rather than the builtin. That said I didn’t double check how thorough the tests are...

**Comment by mattjj** [comment]
Thanks for the thorough review!

File: jax/lax/lax.py (hunk 1 above)
Comment (mattjj) [reply jax/lax/lax.py]: I double-checked that we use `onp.shape` in other places in lax.py.

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (mattjj) [reply jax/numpy/lax_numpy.py]: Looks to me like the tests cover this!

File: tests/lax_numpy_test.py (hunk 3 above)
Comment (mattjj) [reply tests/lax_numpy_test.py]: Done.
